### PR TITLE
Remap msys2 usb, fix USB port detection

### PR DESCRIPTION
**Pull request by nooges**
- Maps tty device in MSYS2 to correct COM port for use with avrdude. (Ex. /dev/ttyS2 -> /dev/COM3, /dev/ttyS0 -> /dev/COM1)
- Fix detection of USB port

### CODE REVIEW

**Comment by drashna** [comment]
Tested this out, it still errors out for me (MSYS2, Windows 10),

> avrdude.exe: ser_open(): can't open device "/??/COM10": The filename, directory name, or volume label syntax is incorrect.

**Comment by nooges** [comment]
@drashna Can you send me the output of /proc/version in your setup?

**Comment by drashna** [comment]
`/proc/version` doesn't work for me. 
However: 
`gcc.exe (Rev1, Built by MSYS2 project) 7.2.0`
And it looks like "20161025" is the build date for it.

**Comment by nooges** [comment]
Okay, fixed things up

**Comment by nooges** [comment]
@drashna Can you test this out again?

**Comment by drashna** [comment]
np. 
Now it errors out doing this: 

> Detecting USB port, reset your controller now.......
> Detected controller on USB port at /dev/ttyS9
> Remapped MSYS2 USB port to /dev/COM10
> avrdude.exe: ser_open(): can't open device "C:/msys64/dev/COM10": The system cannot find the file specified.
> 
> 
> avrdude.exe done.  Thank you.
> 
>
>make[1]: *** [tmk_core/avr.mk:172: avrdude] Error 1

The COM port is correct (COM10), but ... the command is apparently prepending the MSYS path.

**Comment by nooges** [comment]
Yeah, the AVRDUDE binary is only available in the MINGW64 shell. You might want to try updating your version of MSYS, most people I've talked to have this version in /proc/version:
`MINGW64_NT-6.1 version 2.9.0(0.318/5/3) (Alexx@WARLOCK) (gcc version 6.3.0 (GCC) ) 2017-09-13 23:16`

**Comment by drashna** [comment]
Yeah, that's the version that I have. 
Same behavior in both (and the copy of avrdude is the same, as well)

File: tmk_core/avr.mk [hunk 1]
```diff
@@ -183,6 +183,10 @@ avrdude: $(BUILD_DIR)/$(TARGET).hex check-size
 		done; \
 		echo ""; \
 		echo "Detected controller on USB port at $$USB"; \
+		if grep -q -s 'MINGW\|MSYS' /proc/version; then \
+			USB=`echo "$$USB" | perl -pne 's/ttyS(\d+)/COM.($$1+1)/e'`; \
```
Comment (drashna) [inline tmk_core/avr.mk]: This fixes the flashing issue:

`USB=`echo "COM"$$(($$(echo $$USB | sed -e 's/\/dev\/ttyS\([0-9]*\)/\1/g') + 1))`; \`

Comment (drashna) [reply tmk_core/avr.mk]: This would fix #2099

Comment (nooges) [reply tmk_core/avr.mk]: Cool, I'll test on my machine and make the change if it works.

Comment (drashna) [reply tmk_core/avr.mk]: Well, I tested if before posting it here, but please do make sure it works, as well!

Comment (Tiryoh) [reply tmk_core/avr.mk]: I also checked the script. It might work.
https://github.com/qmk/qmk_firmware/issues/2099#issuecomment-365719393

**Comment by drashna** [comment]
Confirmed that these changes fix the issue with MSYS/Windows, and flash properly.

The only caveat now is that you have avrdude installed somewhere accessible. 
It may be a good idea (eg, I have no idea how) to download the EXE for Windows (MinGW) and place it in the qmk_util directory with some of the other tools (like teensy_loader_cli.exe).

And it looks like the official source/site for that is: 
http://download.savannah.gnu.org/releases/avrdude/avrdude-6.3-mingw32.zip
(this is the version, and source mentioned in the avrdude binary included with the QMK toolkit, as well)

**Comment by jackhumbert** [comment]
Awesome - thanks!

**Pull request merged**